docs/topics/multiplatform/multiplatform-hierarchy.md

```diff
@@ -0,0 +1,108 @@
+
+```kotlin
+kotlin {
+    ios() // iOS device and simulator targets; iosMain and iosTest source sets
+}
+```
+
+The Kotlin toolchain will provide the correct default dependencies and find the API surface area available in the shared
+code. This prevents such cases as, for example, the use of a macOS-specific function in the code shared for Windows.
+
+## For library authors
+
+A hierarchical project structure allows reusing code in similar targets, as well as publishing and consuming libraries
+with granular APIs targeting similar platforms.
+
+The Kotlin toolchain will automatically figure out the API available in the consumer source set while checking the
+unsafe usages, like using an API meant for the JVM in JS code.
+
+* Libraries published with the hierarchical project structure are compatible only with projects that have hierarchical
+  project structure. To enable compatibility with non-hierarchical projects, add the following to
+  the `gradle.properties` file in your library project:
+
+   ```properties
+   kotlin.mpp.enableCompatibilityMetadataVariant=true
+   ```
+
+* Libraries published without the hierarchical project structure can’t be used in a shared native source set. For
+  example, users with `ios()` shortcuts in their `build.gradle.(kts)` files won’t be able to use your library in their
+  iOS-shared code.
+
+See [Compatibility](#compatibility) for more details.
+
+## Compatibility
+
+The compatibility between multiplatform projects and libraries is as follows:
+
+| Library with hierarchical project structure | Project with hierarchical project structure | Compatibility                                            |
+|---------------------------------------------|---------------------------------------------|----------------------------------------------------------|
+| Yes                                         | Yes                                         | ✅                                                        |
+| Yes                                         | No                                          | Need to enable with `enableCompatibilityMetadataVariant` |
+| No                                          | Yes                                         | Library can't be used in a shared native source set      |
+| No                                          | No                                          | ✅                                                        |
+
+## How to opt-out
+
+To disable hierarchical structure support, set the following option to `false` in your `gradle.properties`:
+
+```properties
+kotlin.mpp.hierarchicalStructureSupport=false
+```
+
+As for the `kotlin.mpp.enableCompatibilityMetadataVariant` option that enables compatibility of libraries published with
+the hierarchical project structure and non-hierarchical projects, you should disable it separately:
+
+```properties
+kotlin.mpp.enableCompatibilityMetadataVariant=false
```

Comment:
That might sound a little bit confusing. 
Yes, it is true that the user may have following set of properties:

```
kotlin.mpp.hierarchicalStructureSupport=false
kotlin.mpp.enableCompatibilityMetadataVariant=true
```

with meaning that they still want to compile `commonMain` source set into legacy metadata artifact. Along with new klib variant. But disable hierarchical support for some reason. However by default `enableCompatibilityMetadataVariant` is false already, so there is no logical reason to explicitly opt-out.

Reply:
Fixed the description to a default option